docs: add documentation for current release process (#3037)
coveo/ui-kit: repo-docs/release-process.md

> # Release processes

Author: because I don't want implementers looking at our repo to think this is documentation that may be relevant to them. I'll rename this `internal-docs`.

This repository contains two release processes, which are triggered on Git commits:

## 1. The pre-release process

The purpose of the pre-release process is to publish new changes from the main branch as frequently as possible. This has two main benefits:

1. This tests whether we broke some parts of the release process so we don't get surprises when we need a manual release.
2. This enables implementers to test a new feature or fix before we trigger a manual release.

To achieve its purpose, the pre-release process is executed on every new commit on the main branch. Additionally, pre-releases do not commit anything to the main branch, which allows them to be deployed even if multiple features are merged on the main branch faster than the CI can release them.

> ## 2. The manual release process

Reviewer: Could we rename this to `The weekly release process` or `The scheduled release process` e.g. instead of `manual` to remove any confusion since it is automated now. And then we can refer to it as `the release` or `the scheduled release` in the text below.

Author: It's still almost entirely manual unfortunately. What's triggered weekly prompts us to approve it, and then often still requires intervention later on.

Still, I'm fine with `scheduled release process`, since we do it on a regular schedule, and I think that describes pretty well what users expect. I'll rename it.

The purpose of the manual release process is to deploy versions of our packages that we feel confident are safe for implementers to use.

> To achieve that purpose, the manual release process is triggered manually whenever we feel that the version currently on the main branch is safe. Additionally, every Wednesday, to ensure we don't forget to regularly trigger manual releases, a request for approval to deploy is triggered.

Reviewer: Can we revert the order with automate releases as the primary mechanism and defaulting to manual releases if there is any issue etc.

Whenever a manual release is successful, a commit is pushed to the main branch which bumps the version of every publishable package in the repo.

# Versioning & publishing to NPM

Versions for any given commit are determined based on the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) specification.

> Specifically, a commit will determine its version by looking for the last manual release version and bumping it based on how breaking the changes are between then and the current commit. This ensures that implementers can safely update their dependencies on our packages without unexpectedly causing errors.

Reviewer: Feels like this explains what conventional commit and semantic versioning is. Wouldn't linking to them be sufficient?

When triggered, releases processes will execute a series of [Nx tasks](https://nx.dev/core-features/run-tasks). Some tasks are run at the root of the repository, and some will be run on each individual package.

## `release:phase0` (lock the main branch)

This task is only run for the manual release.

> The purpose of this task is to lock the main branch, preventing users from merging pull requests of committing.

Reviewer: ```suggestion
The purpose of this task is to lock the main branch, preventing users from merging pull requests while the release is in progress.
```

Specifically, this prevents new fixes or features from getting merged between the current release and the version bump commit we're about to push. If fixes or features were merged before the upcoming version bump, they would not be taken into account when calculating versions for the next releases.

This task accomplishes its purpose by:

> 1. Committing a ".git-lock` file to the main branch.

Reviewer 2: ```suggestion
1. Committing a `.git-lock` file to the main branch.
```

> 2. Updating the repository's settings to enable "Require branches to be up to date before merging" for the main branch.

Reviewer: Broader question - why not have this enabled by default ?

Author: For most repos, I think it'd be fine, but our repo is a monorepo with multiple teams working on it and a lot of PRs open at once, and our CI still takes quite a few minutes to run. This would block people very often and be very annoying. We could investigate using merge queues, but I haven't looked much into it.

## `release:phase2` (publish to NPM)

This task is run individually on every package, immediately after `release:build`.

This workflow works on the assumption that we can't bump a package's version if any of its dependencies isn't published to NPM. The purpose of this task is to publish packages to NPM before dependant packages bump their versions.

If a package is already published to NPM, this task will exit without error. After a package is published, this task will repeatedly query NPM until it confirms that the package exists in the registry.

## `release:phase3` (reify `package-lock.json`)

This task is run once at the root of the repository after every package's version was bumped.

The purpose of this task is to re-generate the root `package-lock.json` file using [Arborist](https://www.npmjs.com/package/@npmcli/arborist), which should confirm that the version bumps were successful.

> Manual releases will commit the new `package-lock.json` to the main branch, which ensures that contributors can run `npm install` without unnecessary changes being made to their `package-lock.json`.

Reviewer 3: what happens at this phase at this phase in the pre release ?

> Manual releases will commit the new `package-lock.json` to the main branch, which ensures that contributors can run `npm install` without unnecessary changes being made to their `package-lock.json`.

Reviewer: Isn't `npm ci` the one which installs based on `package-lock.json` without modifying it and `npm install` would read `package.json` and update deps which sometimes would update `package-lock.json`?

Author: Unfortunately, `npm install` also suffered from a broken `package-lock.json` when we used Lerna. We were forced to delete the whole `node_modules` before doing `npm install`, and then we'd get diffs between Lerna's `package-lock.json` and the local one.

## `release:phase4` (commit version bumps)

This task is only run for the manual release.

This task will create a new "version bump" commit, which will contain:

- The new `package.json` files.
- The new `package-lock.json` file.
- The new `CHANGELOG.md` files.
- Git tags for each bumped package version.

This task will also revert the changes from `release:phase0` to allow merging new features and fixes.

# Deploying

Whenever a new version bump is pushed to the main branch by the manual release process, the Jenkins CI will run the `Jenkinsfile` on it.

The purpose of this step is to:

1. Publish some packages to the dev and staging CDN.
2. After QA approval:

>    1. Publish Quantic to Salesforces.

Reviewer 2: ```suggestion
   1. Publish Quantic to Salesforce.
```
